### PR TITLE
Code for Adam Optimizer

### PR DESCRIPTION
**Pull request by terrytangyuan**

### CODE REVIEW

**Comment by javelinjs** [comment]
@terrytangyuan Could you PR example/image-classification/README.md and example/rnn/lstm.py to dmlc/mxnet/master directly? These shouldn't be parts of scala-package

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/Adam.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    val lr =
+      (if (lrScheduler != null) {
+        val scheduledLr = lrScheduler(numUpdate)
+        updateCount(index)
+        scheduledLr
+      } else {
+        this.learningRate
+      }) * lrScale.getOrElse(index, 1f)
+
+    var (mean, variance)  = state
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/Adam.scala]: state.asInstanceOf[(NDArray, NDArray) to restrict the type of mean and variance here

**Comment by terrytangyuan** [comment]
Those are recent changes I pulled from dmlc/mxnet. It doesn't count as my commit. Sorry for the confusion. They will not be in scala-package. I'll make sure not to include them here next time.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/Adam.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    val beta1t = beta1 * math.pow(decayFactor, t1 - 1) toFloat
+
+    var resdGrad = grad * rescaleGrad
+    if (clipGradient != 0f) {
+      resdGrad = NDArray.clip(resdGrad, -clipGradient, clipGradient)
+    }
+
+    val meanT = beta1t * mean.asInstanceOf[NDArray] + (1.0 - beta1t) * resdGrad toScalar
+    val varianceT = beta2 * variance.asInstanceOf[NDArray] + (1.0f - beta2) * resdGrad * resdGrad toScalar
+
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/Adam.scala]: why toScalar here?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/Adam.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    var step = learningRate * meanT / (math.sqrt(varianceT) + epsilon)
+
+    if (wd > 0.0f) {
+      step += (lr * wd * weight).toScalar
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/Adam.scala]: why toScalar?

**Comment by javelinjs** [comment]
Oh I see. I suggest to always pull from dmlc on master branch.

**Pull request merged**